### PR TITLE
datapath: Fix lo and fallback node address selection

### PR DESCRIPTION
**Pull request by joamaki**
```
datapath: Pick non-loopback addresses assigned to lo as node addresses

    This fixes a regression where the non-loopback addresses assigned to the
    loopback device (lo) were not considered host/node addresses. This broke the
    practice of assigning VIPs to the loopback device to make Cilium consider
    them.

    The problem was due to filtering on "ExcludedDevicePrefixes" that included
    "lo". This filtering is already done in the devices controller that
    populates the Table[Device] that this code reads, so this filtering can be
    dropped. In addition to this the fix whitelists the lo device to make it
    unnecessary to specify "--devices=lo,..." and thus retaining the same
    semantics as v1.14.

    Fixes: #33214

    Signed-off-by: Jussi Maki <jussi@isovalent.com>

datapath: Skip lxc devices for fallback and prefer selected

    The fallback is used for e.g. BPF masquerading when the target device has no
    address, this is "best effort" for ECMP etc. setups). The selection
    algorithm for the fallback node address wasn't taking into account whether
    the device was selected or not, which led to the fallback address being
    taken from non-selected devices.

    Add selected as the first criteria for checking if a fallback is better. And
    to avoid unnecessary churn on updating the fallback, always skip lxc*
    devices.

    Signed-off-by: Jussi Maki <jussi@isovalent.com>
```

```release-note
Fix issue in picking node IP addresses from the loopback device. This fixes a regression in v1.15 and v1.16 where VIPs assigned to the lo device were not considered by Cilium.
Fix spurious updates node addresses to avoid unnecessary datapath reinitializations.
```

### CODE REVIEW

**Review by borkmann — APPROVED** [review]
lgtm 🚀

**Review by dylandreimerink — APPROVED** [review]
LGTM!

**Review by markpash — APPROVED** [review]

File: pkg/datapath/tables/node_address.go [hunk 1]
```diff
@@ -383,6 +388,10 @@ func (n *nodeAddressController) updateFallbacks(txn statedb.ReadTxn, dev *Device
 		fallbacks.clear()
 		devices := n.Devices.All(txn)
 		for dev, _, ok := devices.Next(); ok; dev, _, ok = devices.Next() {
+			if strings.HasPrefix(dev.Name, "lxc") {
```
Comment (tommyp1ckles) [inline pkg/datapath/tables/node_address.go:391]: question: would "lxc_" be a more specific prefix to look for?

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/datapath/tables/node_address.go (hunk 1 above)
Comment (joamaki) [reply pkg/datapath/tables/node_address.go:391]: No:
```
root@kind-worker:/home/cilium# cilium statedb devices
# Name            Index   Selected   Type     MTU     HWAddr              Flags                    Addresses
cilium_vxlan      4       false      vxlan    1500    46:1b:fc:10:41:ea   up|broadcast|multicast   fe80::441b:fcff:fe10:41ea
lxcd88bffc8e0a6   8       false      veth     1500    a2:3b:c8:f2:dc:16   up|broadcast|multicast   fe80::a03b:c8ff:fef2:dc16
lxc899f9560ba1b   10      false      veth     1500    ba:9e:f8:bc:ab:a7   up|broadcast|multicast   fe80::b89e:f8ff:febc:aba7
...   
```

**Pull request merged**